pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,188 @@
+package logger
+
+import (
+	"fmt"
+	"go.uber.org/zap/zapcore"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+type ECSErrorLogger struct {
+	ErrorFilePath    string
+	ErrorLogMaxAge   int
+	ErrorFileMaxSize int
+
+	queue     []zapcore.Entry
+	ticker    time.Ticker
+	errorFile *os.File
+	size      int
+}
+
+// ECSErrorReporter func help select the different case between either write error log into file or
+// rotate and delete old logs in file based on TTL
+func (l *ECSErrorLogger) ECSErrorReporter() {
+	defer l.ticker.Stop()
+	for {
+		select {
+		// when new error add in the channel, this case will be selected and write it into error log file
+		case e := <-errorLogChannel:
+			l.Write(e)
+		// this case will run every x mins (x is configurable) to delete the error which already expired, error log max age
+		// depends on STATUS_MESSAGE_TTL
+		case <-l.ticker.C:
+			l.processTimeout()
+		}
+	}
+}
+
+func (l *ECSErrorLogger) Write(newError zapcore.Entry) {
+	if l.errorFile == nil {
+		if err := l.openExistingOrNewErrorFile(); err != nil {
+			log.Printf("could not open error log file when write new error, err: %v", err)
+		}
+	}
+	newErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newError.Time, newError.Level, newError.Caller, newError.Message))
+
+	if len(newErrorLog) > l.ErrorFileMaxSize {
+		log.Printf("error size exceed the max size of error file")
+		return
+	}
+
+	err := l.errorFile.Truncate(0)
+	if err != nil {
+		log.Printf("could not truncate error log file when write new error, err: %v", err)
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Message != newError.Message {
+			n, err := l.errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("could not write error into error log file, err: %v", err)
+			}
+			l.size += n
+			l.queue = append(l.queue, e)
+		}
+		l.queue = l.queue[1:]
+	}
+	// delete the old errors if old errors plus new error exceed the max size
+	if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+		l.rotate(newErrorLog)
+	}
```

Comment:
So as we talked about on Wednesday, I think it's best if you check the size before you write the file. Why write all errors to the file first and then check size second?

Why shouldn't the algorithm be:
```
create a message queue;
 when (receive a new error log message) {
    calculate all message size;
    truncate the message to 1KB first;
    
    while (all message size > 1KB) {
        remove the oldest message;
    }
    
    if (message is already in message queue) {
        remove old message
    } 
    add new message into message queue;
    
    delete the local file;
    write all messages in message queue to local file
}
```

Reply:
because there is a case that an old error is same as the new one and file size plus new error size exceed the max limit, and if we iterate the queue first, we don't need to write the old one into the file and the new one will directly add to the file.
if an old error is same as the new one and we check the total size first, it will count the same error size twice and delete another error from file.
is this logic correct?

Reply:
I see. Good point. But still you can/should do the write to file last.

I think you are saying it should be:

```
create a message queue;
 when (receive a new error log message) {
    calculate all message size;
    truncate the message to 1KB first;

    if (message is already in message queue) {
        remove old message
    } 

   add new message into message queue;
    
    while (all message size > 1KB) {
        remove the oldest message;
    }
    
    delete the local file;
    write all messages in message queue to local file
}
```